### DISCUSSION TITLE
Mutable Managed State

### DISCUSSION THREAD

**Discussion by james-d-hasselman** [tauri-apps/tauri#3911] [Q&A] [answered]
Is it possible to mutate managed state within a Tauri Command?

I have the following function:

```
#[tauri::command]
fn initialize_interpreter(
    app_handle: tauri::AppHandle,
    window: tauri::Window,
    interpreter: State<Option<Interpreter>>,
    rom: Rom,
) {
    let mut rom_file = File::open(rom.path).unwrap();
    let mut rom = vec![];
    rom_file.read_to_end(&mut rom).unwrap();
    let display = TauriDisplay::new(window.clone());
    let keyboard = TauriKeyboard::new(app_handle.clone());
    let buzzer = JavaScriptAudio::new(window.clone());
    let mut interpreter = interpreter;
    *interpreter = Some(Interpreter::new(Box::new(display), Box::new(buzzer), Box::new(keyboard), &rom));
}
```

 which does not compile because "trait `DerefMut` is required to modify through a dereference, but it is not implemented for `tauri::State<'_, std::option::Option<Interpreter>>`"
 
 I'm not really sure what to do here. I have tried wrapping the Option in both Box and Arc thinking that would solve the problem, but apparently it doesn't.

Thanks.

**Comment by FabianLars** [comment, marked as answer]
You need to use a container that enables _interior mutability_, like a [Mutex](https://doc.rust-lang.org/std/sync/struct.Mutex.html). Here's an example in the context of a tauri app: https://github.com/tauri-apps/tauri/blob/dev/examples/state/main.rs#L32

**Reply by james-d-hasselman** [reply]
Thank you for the prompt response. I had checked the docs but didn't think to look at the examples. I will remember that next time.
